// react-bootstrap-typeahead/types/components/Overlay.js
import PropTypes from 'prop-types';
import { useState } from 'react';
import { usePopper } from 'react-popper';
import { noop } from '../utils';
export var Align;
(function (Align) {
    Align["JUSTIFY"] = "justify";
    Align["LEFT"] = "left";
    Align["RIGHT"] = "right";
})(Align || (Align = {}));
const SafeElement = typeof Element === 'undefined' ? noop : Element;
const propTypes = {
    align: PropTypes.oneOf(Object.values(Align)),
    children: PropTypes.func.isRequired,
    dropup: PropTypes.bool,
    flip: PropTypes.bool,
    isMenuShown: PropTypes.bool,
    positionFixed: PropTypes.bool,
    referenceElement: PropTypes.instanceOf(SafeElement),
};
const defaultProps = {
    align: Align.JUSTIFY,
    dropup: false,
    flip: false,
    isMenuShown: false,
    positionFixed: false,
};
const setPopperWidth = {
    enabled: true,
    fn: (data) => {
        data.state.styles.popper.width = `${data.state.rects.reference.width}px`;
    },
    name: 'setPopperWidth',
    phase: 'write',
};
export function getModifiers(props) {
    const modifiers = [
        {
            enabled: !!props.flip,
            name: 'flip',
        },
    ];
    if (props.align !== Align.RIGHT && props.align !== Align.LEFT) {
        modifiers.push(setPopperWidth);
    }
    return modifiers;
}
export function getPlacement(props) {
    const x = props.align === Align.RIGHT ? 'end' : 'start';
    const y = props.dropup ? 'top' : 'bottom';
    return `${y}-${x}`;
}
const Overlay = ({ referenceElement, ...props }) => {
    const [popperElement, attachRef] = useState();
    const { attributes, styles, forceUpdate } = usePopper(referenceElement, popperElement, {
        modifiers: getModifiers(props),
        placement: getPlacement(props),
        strategy: props.positionFixed ? 'fixed' : 'absolute',
    });
    if (!props.isMenuShown) {
        return null;
    }
    return props.children({
        ...attributes.popper,
        innerRef: attachRef,
        inputHeight: referenceElement?.offsetHeight || 0,
        scheduleUpdate: forceUpdate,
        style: styles.popper,
    });
};
Overlay.propTypes = propTypes;
Overlay.defaultProps = defaultProps;
export default Overlay;
